sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.plans.logical.statsEstimation.FilterEstimation
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * A rule that reorder conditions in filters base on estimated selectivity and compute cost.
+ *
+ * {{{
+ *   SELECT * FROM lineitem WHERE l_comment LIKE '%a%' AND l_orderkey = 1024 ==>
+ *   SELECT * FROM lineitem WHERE l_orderkey = 1024 AND l_comment LIKE '%a%'
+ * }}}
+ */
+object PredicateReorder extends Rule[LogicalPlan] with PredicateHelper {
+
+  private val DEFAULT_SELECTIVITY = 0.1D
```

Comment:
This default value from Impala: https://github.com/apache/impala/blob/d271baa33da1a02aa6ffc47b0380dc62239107b4/fe/src/main/java/org/apache/impala/analysis/Expr.java#L76-L77

May be we can improve the default selectivity estimate in the future:
https://github.com/postgres/postgres/blob/ed934d4fa30f0f94e6f7125ad2154e6a58d1c7f7/src/include/utils/selfuncs.h#L33-L49